### PR TITLE
Create a typings file per source file, instead of per package.

### PR DESCRIPTION
**Pull request by aomarks**
This would allow you to pull in typings only for the parts of a package you are using.

### CODE REVIEW

File: scripts/make-goldens.js [hunk 1]
```diff
@@ -20,20 +20,23 @@
...
-const rimraf = require('rimraf');
+
 const {generateDeclarations} = require('../lib/gen-ts');
 
 const fixturesDir = path.join(__dirname, '..', 'src', 'test', 'fixtures');
 const goldensDir = path.join(__dirname, '..', 'src', 'test', 'goldens');
 
-rimraf.sync(goldensDir);
-fs.mkdirSync(goldensDir);
+fsExtra.emptyDir(goldensDir);
```
Comment (rictic) [inline scripts/make-goldens.js:31]: Why this change?

File: src/cli.ts [hunk 2]
```diff
@@ -61,15 +63,27 @@ async function run(argv: string[]) {
...
+    await writeFileMap(args.outDir, fileMap);
   } else {
-    process.stdout.write(declarations);
+    const concatenated = [...fileMap.values()].join('\n');
+    process.stdout.write(concatenated);
   }
 }
 
+function writeFileMap(
+    rootDir: string, files: Map<string, string>): Promise<void[]> {
```
Comment (rictic) [inline src/cli.ts]: It's a little weird to return a promise of an array of undefineds. Recommend instead returning just Promise<void> by awaiting the Promise.all rather than returning it.

File: src/gen-ts.ts [hunk 3]
```diff
@@ -18,37 +19,55 @@ import {serializeTsDeclarations} from './ts-serialize';
...
+function analyzerToAst(analysis: analyzer.Analysis): ts.Document[] {
+  const tsDocs = new Array<ts.Document>();
+  for (const analyzerDoc of analysis.getFeatures({kind: 'js-document'})) {
     // TODO This is a very crude exclusion rule.
-    if (doc.url.match(/(test|demo)\//)) {
+    if (analyzerDoc.url.match(/(test|demo)\//)) {
       continue;
     }
-    handleDocument(doc, root);
+    // TODO I think there can be many analyzer documents with the same URL,
```
Comment (rictic) [inline src/gen-ts.ts]: There can be at most one toplevel document at a URL, but yeah there can be multiple inline documents, and they should be merged here I think

File: src/test/gen-ts_test.ts [hunk 4]
```diff
@@ -21,11 +22,21 @@ const goldens = path.join(__dirname, '..', '..', 'src', 'test', 'goldens');
 suite('generateDeclarations', () => {
   for (const fixture of fs.readdirSync(goldens)) {
     test(fixture, async () => {
-      const golden =
-          fs.readFileSync(path.join(goldens, fixture, 'expected.d.ts'));
-      const declarations =
-          await generateDeclarations(path.join(fixtures, fixture));
-      assert.equal(declarations, golden.toString());
+      // TODO Make these tests more async.
```
Comment (rictic) [inline src/test/gen-ts_test.ts]: More async in what way?

File: src/test/gen-ts_test.ts [hunk 5]
```diff
@@ -21,11 +22,21 @@ const goldens = path.join(__dirname, '..', '..', 'src', 'test', 'goldens');
...
+      const actual = await generateDeclarations(path.join(fixtures, fixture));
+      const golden = readDirAsMap(path.join(goldens, fixture));
+      (assert as any).hasAllKeys(actual, [...golden.keys()]);
+      for (const filename of actual.keys()) {
+        assert.equal(actual.get(filename), golden.get(filename), filename);
+      }
     }).timeout(30000);  // These tests can take a long time.
```
Comment (rictic) [inline src/test/gen-ts_test.ts:31]: Wow, how long do these tests usually take?

File: scripts/make-goldens.js (hunk 1 above)
Comment (aomarks) [reply scripts/make-goldens.js:31]: I added a dependency on `fs-extra` because I wanted `mkdirs` and `outputFile`. Since `fs-extra` includes `emptyDir`, which is equivalent to `rimraf` followed by `fs.mkdir`, I used that instead so that I could drop the dependency on `rimraf`.

File: src/cli.ts (hunk 2 above)
Comment (aomarks) [reply src/cli.ts]: Done.

File: src/test/gen-ts_test.ts (hunk 4 above)
Comment (aomarks) [reply src/test/gen-ts_test.ts]: Reading all the golden files are done synchronously. They could be done concurrently. The bottleneck by far is analyzer though, so I'm not bothered right now, as it makes the code harder to read. Removed the TODO.

File: src/test/gen-ts_test.ts (hunk 5 above)
Comment (aomarks) [reply src/test/gen-ts_test.ts:31]: 2-3 seconds, but occasionally longer. Travis is sometimes resource constrained so I wanted to give a fairly big buffer.

File: src/gen-ts.ts (hunk 3 above)
Comment (aomarks) [reply src/gen-ts.ts]: Done.

**Comment by aomarks** [comment]
All changes done. PTAL!

**Comment by aomarks** [comment]
Fixed the commits, latest commit is now just changes since your review.

**Review by rictic — APPROVED** [review]
